python/cog/server/worker.py

```diff
@@ -480,20 +501,19 @@ async def _aloop(
         task = None
 
         with scope(self._loop_scope()), redirector:
-            while True:
-                e = cast(Envelope, await self._events.recv())
-                if isinstance(e.event, Cancel) and task and self._cancelable:
-                    task.cancel()
-                elif isinstance(e.event, Shutdown):
-                    break
-                elif isinstance(e.event, PredictionInput):
-                    task = asyncio.create_task(
-                        self._apredict(e.tag, e.event.payload, predict, redirector)
-                    )
-                else:
-                    print(f"Got unexpected event: {e.event}", file=sys.stderr)
-            if task:
-                await task
+            async with asyncio.TaskGroup() as tg:
+                while True:
+                    e = cast(Envelope, await self._events.recv())
+                    if isinstance(e.event, Cancel) and task and self._cancelable:
+                        task.cancel()
+                    elif isinstance(e.event, Shutdown):
+                        break
+                    elif isinstance(e.event, PredictionInput):
+                        task = tg.create_task(
```

Comment:
Does this potentially memory leak all these events forever by adding them to the task group?

Reply:
Completed tasks are removed from the task group once they resolve. So no, this wont leak memory. However if any task raises an exception other than an [`asyncio.CancelledError`](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError) the existing tasks in the group will be cancelled. I don't think this is what we want…